File: backend/src/utils/splitCalculator.js
/**
 * Compute each participant's share of an expense.
 * mode: 'equal' | 'exact' | 'percent' | 'shares'
 * splits: [{ userId, value? }]
 *   - equal: value ignored
 *   - exact: value = amount (must sum to total)
 *   - percent: value = 0-100 (must sum to 100)
 *   - shares: value = positive weight
 *
 * Returns: [{ userId, amount }] (rounded to 2dp, residual goes to first participant)
 */
import { BadRequest } from './errors.js';

export function computeShares({ total, mode, splits }) {
  if (!Array.isArray(splits) || splits.length === 0) {
    throw BadRequest('Pick at least one participant to split with', 'SPLIT_EMPTY');
  }
  const round = (n) => Math.round(n * 100) / 100;
  let shares;

  if (mode === 'equal') {
    const each = total / splits.length;
    shares = splits.map((s) => ({ userId: s.userId, amount: round(each) }));
  } else if (mode === 'exact') {
    const sum = splits.reduce((a, s) => a + Number(s.value || 0), 0);
    if (Math.abs(sum - total) > 0.01) {
      throw BadRequest(
        `Exact amounts must add up to the total (${round(sum)} of ${round(total)})`,
        'SPLIT_EXACT_MISMATCH',
      );
    }
    shares = splits.map((s) => ({ userId: s.userId, amount: round(Number(s.value)) }));
  } else if (mode === 'percent') {
    const sum = splits.reduce((a, s) => a + Number(s.value || 0), 0);
    if (Math.abs(sum - 100) > 0.01) {
      throw BadRequest(
        `Percentages must add up to 100% (currently ${round(sum)}%)`,
        'SPLIT_PERCENT_MISMATCH',
      );
    }
    shares = splits.map((s) => ({ userId: s.userId, amount: round((Number(s.value) / 100) * total) }));
  } else if (mode === 'shares') {
    const sum = splits.reduce((a, s) => a + Number(s.value || 0), 0);
    if (sum <= 0) throw BadRequest('Share weights must be greater than zero', 'SPLIT_SHARES_INVALID');
    shares = splits.map((s) => ({ userId: s.userId, amount: round((Number(s.value) / sum) * total) }));
  } else {
    throw BadRequest(`Unknown split mode: ${mode}`, 'SPLIT_MODE_INVALID');
  }

  // Reconcile rounding residual onto the first participant
  const sumShares = shares.reduce((a, s) => a + s.amount, 0);
  const residual = round(total - sumShares);
  if (Math.abs(residual) >= 0.01) {
    shares[0].amount = round(shares[0].amount + residual);
  }
  return shares;
}

File: app/lib/core/balances/split_calculator.dart
/// Dart port of the backend `splitCalculator` + payer resolution, so an expense
/// added offline produces the same shares/payers locally as the server would.

class ShareResult {
  ShareResult(this.userId, this.amount);
  final String userId;
  final double amount;
}

double _round2(double n) => (n * 100).round() / 100;

/// Returns each participant's share. `splits` is `[{userId, value?}]`.
/// Throws [ArgumentError] with a friendly message on invalid input (mirrors the
/// server's BadRequest messages so offline validation matches online).
List<ShareResult> computeShares({
  required double total,
  required String mode,
  required List<Map<String, dynamic>> splits,
}) {
  if (splits.isEmpty) {
    throw ArgumentError('Pick at least one participant to split with');
  }
  List<ShareResult> shares;
  double valueOf(Map<String, dynamic> s) => (s['value'] as num?)?.toDouble() ?? 0;

  switch (mode) {
    case 'equal':
      final each = total / splits.length;
      shares = [for (final s in splits) ShareResult(s['userId'].toString(), _round2(each))];
      break;
    case 'exact':
      final sum = splits.fold<double>(0, (a, s) => a + valueOf(s));
      if ((sum - total).abs() > 0.01) {
        throw ArgumentError(
            'Exact amounts must add up to the total (${_round2(sum)} of ${_round2(total)})');
      }
      shares = [for (final s in splits) ShareResult(s['userId'].toString(), _round2(valueOf(s)))];
      break;
    case 'percent':
      final sum = splits.fold<double>(0, (a, s) => a + valueOf(s));
      if ((sum - 100).abs() > 0.01) {
        throw ArgumentError('Percentages must add up to 100% (currently ${_round2(sum)}%)');
      }
      shares = [
        for (final s in splits) ShareResult(s['userId'].toString(), _round2(valueOf(s) / 100 * total))
      ];
      break;
    case 'shares':
      final sum = splits.fold<double>(0, (a, s) => a + valueOf(s));
      if (sum <= 0) throw ArgumentError('Share weights must be greater than zero');
      shares = [
        for (final s in splits) ShareResult(s['userId'].toString(), _round2(valueOf(s) / sum * total))
      ];
      break;
    default:
      throw ArgumentError('Unknown split mode: $mode');
  }

  // Reconcile rounding residual onto the first participant.
  final sumShares = shares.fold<double>(0, (a, s) => a + s.amount);
  final residual = _round2(total - sumShares);
  if (residual.abs() >= 0.01) {
    shares[0] = ShareResult(shares[0].userId, _round2(shares[0].amount + residual));
  }
  return shares;
}

/// Normalises payer input into `(paidBy, payers)`. `payers` is empty for a
/// single-payer expense. Throws [ArgumentError] if multi-payer amounts don't
/// add up to [total].
({String paidBy, List<ShareResult> payers}) resolvePayers({
  required String paidBy,
  required List<Map<String, dynamic>> payers,
  required double total,
}) {
  final nonZero = payers.where((p) => ((p['amount'] as num?)?.toDouble() ?? 0) > 0).toList();
  if (nonZero.length <= 1) {
    final single = nonZero.isNotEmpty ? nonZero.first['userId'].toString() : paidBy;
    return (paidBy: single, payers: <ShareResult>[]);
  }
  final sum = nonZero.fold<double>(0, (a, p) => a + ((p['amount'] as num).toDouble()));
  if ((sum - total).abs() > 0.01) {
    throw ArgumentError('Payer amounts must add up to the total (${_round2(sum)} of ${_round2(total)})');
  }
  final list = [for (final p in nonZero) ShareResult(p['userId'].toString(), _round2((p['amount'] as num).toDouble()))];
  final primary = list.reduce((a, b) => b.amount > a.amount ? b : a);
  return (paidBy: primary.userId, payers: list);
}
